apps/prairielearn/src/ee/lib/ai-grading.ts

```diff
@@ -34,165 +30,150 @@ import * as questionServers from '../../question-servers/index.js';
 import { createEmbedding, vectorToString } from './contextEmbeddings.js';
 
 const sql = loadSqlEquiv(import.meta.url);
+const OPEN_AI_MODEL = 'gpt-4o-2024-08-06';
 
 const SubmissionVariantSchema = z.object({
   variant: VariantSchema,
   submission: SubmissionSchema,
 });
-const GPTGradeSchema = z.object({ grade: z.number(), feedback: z.string() });
+const GPTScoreSchema = z.object({ score: z.number(), feedback: z.string() });
 const GradedExampleSchema = z.object({
   submission_text: z.string(),
   score_perc: z.number(),
+  feedback: z.record(z.string(), z.any()).nullable(),
   instance_question_id: z.string(),
+  manual_rubric_grading_id: z.string().nullable(),
 });
 type GradedExample = z.infer<typeof GradedExampleSchema>;
 
-async function generateGPTPromptWithExamples({
-  question_prompt,
-  student_answer,
+async function generateGPTPrompt({
+  questionPrompt,
+  submission_text,
   example_submissions,
+  rubric_items,
 }: {
-  question_prompt: string;
-  student_answer: string;
+  questionPrompt: string;
+  submission_text: string;
   example_submissions: GradedExample[];
-}): Promise<
-  {
+  rubric_items: RubricItem[];
+}): Promise<{
+  messages: {
     role: 'system' | 'user';
     content: string;
-  }[]
-> {
+  }[];
+  warning: string;
+}> {
   const messages: {
     role: 'system' | 'user';
     content: string;
   }[] = [];
+  let warning = '';
 
   // Instructions for grading
-  messages.push({
-    role: 'system',
-    content:
-      'You are an instructor for a course, and you are grading assignments. You should always return the grade using a json object of 2 parameters: grade and feedback. The grade should be an integer between 0 and 100. 0 being the lowest and 100 being the highest, and the feedback should be why you give this grade, or how to improve the answer. You can say correct or leave blank when the grade is close to 100. I will provide some example answers and their corresponding grades.',
-  });
+  if (rubric_items.length > 0) {
+    let rubric_info = '';
+    for (const item of rubric_items) {
+      rubric_info += `description: ${item.description}\n`;
+      if (item.explanation) {
+        rubric_info += `explanation: ${item.explanation}\n`;
+      }
+      if (item.grader_note) {
+        rubric_info += `grader note: ${item.grader_note}\n`;
+      }
+      rubric_info += '\n';
+    }
+    messages.push({
+      role: 'system',
+      content:
+        "You are an instructor for a course, and you are grading assignments. You are provided several rubric items with a description, explanation, and grader note. You must grade the assignment by using the rubric and returning an object of rubric descriptions and whether or not that rubric item applies to the student's submission. If no rubric items apply, do not select any. I will provide some example answers and their corresponding selected rubric items.",
```

Comment:
Do we correctly format the prompt when there aren't any examples available? At the least we might want to leave off the final sentence above?